diff --git a/cmake/Modules/FindQwt.cmake b/cmake/Modules/FindQwt.cmake
index f2cf344..6230025 100644
--- a/cmake/Modules/FindQwt.cmake
+++ b/cmake/Modules/FindQwt.cmake
@@ -10,6 +10,7 @@ find_path (QWT_INCLUDE_DIRS
 	/usr/local/include/qwt
 	/usr/include/qwt-qt4
 	/usr/include/qwt
+	/usr/include/qwt5
 	/opt/local/include/qwt
 	/sw/include/qwt
 )
diff --git a/res/CMakeLists.txt b/res/CMakeLists.txt
index 845c8be..62a5bce 100644
--- a/res/CMakeLists.txt
+++ b/res/CMakeLists.txt
@@ -37,10 +37,6 @@ if(NOT PYUIC4_FOUND)
 endif()
 
 find_package(Qwt)
-if(NOT QWT_FOUND)
-    message(STATUS "Qwt not found, not installing GUI application")
-    return()
-endif()
 
 set(RX_UI_SRC ${CMAKE_CURRENT_SOURCE_DIR}/modes_rx.ui)
 set(RX_UI_PY_PRE_MASSAGE ${CMAKE_CURRENT_BINARY_DIR}/modes_rx_ui_borked.py)
